edx-solutions/cs_comments_service
# [MCKIN-8918] Backport retirement API

This PR backports the user retirement API from Hawthorne and related improvement commits.
The backported commits are:
 * [Add ability to change author_username in content.](https://github.com/edx/cs_comments_service/commit/42fae9159970b3a50e109ee3d37d97f4f86b9287)
 * [Add user retirement API endpoint to remove forum post content.](https://github.com/edx/cs_comments_service/commit/691ac62846ef526bf51dde49ab3abc265e8c09ee)
 * [Shift from deprecated should to expect syntax.](https://github.com/edx/cs_comments_service/commit/24fd748807bbbcc731905746adac024b5e9336a0)
 * [Handle non-existent user and missing param upon retiring user.](https://github.com/edx/cs_comments_service/commit/078c1d3b9f9d70694c2b6a4965388c7a57dbdc8b)
 * [Blank out comment thread documents as well as comment documents.](https://github.com/edx/cs_comments_service/commit/c0c718141a3558a974b634a756d2aabc796b676e)
 * [Add error handling around Elasticsearch update/delete.](https://github.com/edx/cs_comments_service/commit/0e39704298e86e3507b6933762e557e16641dda1)
 * [Process ES updates in bulk during user retirement.](https://github.com/edx/cs_comments_service/commit/7eae611dc7daf295d9e6ca9f48f81903497ba26d)
 * [Make sure ES stays enabled, even if there is an error during update().](https://github.com/edx/cs_comments_service/commit/a8188ad799607e4e4b4c162278ded7c59a96845b)
 * [Avoid sending empty ES bulk requests.](https://github.com/edx/cs_comments_service/commit/6d6770e9ab59fdbadb014f776538d3c57a43ab57)

All of these commits are the implementation of the retirement feature or a improvement related to it.

**Testing:**
1. On your solutions devstack, pull this branch.
2. Create a new user on LMS and enroll in the demonstration course.
3. Create a new discussion and interact with some posts.
4. Delete the user with the backported deletion API
```
curl http://localhost:18080/api/v1/users/1530/retire?retired_username=deleted -X POST -d "" 
```
5. Log in as another user in the LMS and check if the posts and comments were deleted.
6. Inside the vagrant machine of the devstack, check that the mongo table contents have either been deleted on anonymised.
```
$ mongo

> use cs_comments_service
switched to db cs_comments_service

> db.users.find()
...
{ "_id" : "1530", "default_sort_key" : "activity", "external_id" : "1530", "username" : "deleted", "read_states" : [ ], "email" : "" } # Check that username was correctly anonymised

> db.contents.find()
... # You should see a lot of comments with "body"="[deleted]" and "author_username"="deleted".
{ "_id" : ObjectId("5c45ab876362396111525759"), "votes" : { "up" : [ ], "down" : [ ], "up_count" : 0, "down_count" : 0, "count" : 0, "point" : 0 }, "visible" : true, "abuse_flaggers" : [ ], "historical_abuse_flaggers" : [ ], "_type" : "Comment", "parent_ids" : [ ], "at_position_list" : [ ], "body" : "[deleted]", "course_id" : "course-v1:edX+DemoX+Demo_Course", "endorsed" : false, "anonymous" : false, "anonymous_to_peers" : false, "author_id" : "1530", "comment_thread_id" : ObjectId("5c45ab806362396111525757"), "child_count" : 0, "depth" : 0, "author_username" : "deleted", "sk" : "5c45ab876362396111525759", "updated_at" : ISODate("2019-01-21T11:26:01.543Z"), "created_at" : ISODate("2019-01-21T11:22:47.555Z"), "retired_username" : "deleted" }
...
```
**Reviewers:**
